Remove the toolchain in riscv-tools

palmer-dabbelt opened the pull request:
The toolchain pointed to by riscv-tools hasn't been maintained for more
than a year now.  This pointer was necessary when we had out-of-tree
ports of the various toolchain components, but now that everything is
upstream toolchains can be obtained via dozens of standard mechanisms.
Having a broken toolchain so publicly visible causes everyone headaches,
so unless someone wants to step up and maintain this I'm going to remove
it.

palmer-dabbelt committed:
Remove the toolchain in riscv-tools

The toolchain pointed to by riscv-tools hasn't been maintained for more
than a year now.  This pointer was necessary when we had out-of-tree
ports of the various toolchain components, but now that everything is
upstream toolchains can be obtained via dozens of standard mechanisms.
Having a broken toolchain so publicly visible causes everyone headaches,
so unless someone wants to step up and maintain this I'm going to remove
it.

In order to keep the Travis running, I've gone ahead and used
crosstool-ng to build a toolchain.  This should be a lot more reliable
than the old riscv-gnu-toolchain based approach.

palmer-dabbelt commented:
I'm making a sw-dev post and will let this sit until I remember it again.

jim-wilson commented:
You should check to see if this breaks the rocket-chip repo.

aswaterman commented:
@jim-wilson it should be fine.  It won't break rocket-chip immediately, because it points to an older commit hash.  We can fix rocket-chip independently of this commit, or maintain a rocket-chip branch of riscv-tools that has the old behavior.

aswaterman commented:
However, this repo's CI will fail because it needs a RISC-V compiler :)

RobertHenry6bev commented on README.md:
```diff
@@ -16,23 +16,27 @@ the host machine
+* Your favorite Linux distribution may already have packages that
+  include a RISC-V cross compiler, which is probably the fastest way to
+  get started.  As of writing this README (March, 2019) I can trivially
+  find partkages for ALT Linux, Arch Linux, Debian, Fedora, FreeBSD,
```
packages, not partkages

emaste commented on README.md:
```diff
@@ -16,23 +16,27 @@ the host machine
+already have packages for these upstream tools, but if it doesn't then
+here are a handful of my favorites:
+
+* Your favorite Linux distribution may already have packages that
```
Your favorite Linux distribution or BSD may already have...

palmer-dabbelt commented:
I just updated the PR to build a cross compiler via crosstool-ng, if that works then I'll merge this.

aswaterman committed:
Add missing help2man package

aswaterman committed:
Back out of crosstools directory after building it

aswaterman committed:
Don't set CC, which crosstool-ng complains about

aswaterman commented:
I fixed a few problems in the .travis.yml and now it's building...

aswaterman committed:
Add travis_wait directive

aswaterman commented:
@palmer-dabbelt Still working on this.  I've noticed that crosstool's attempt to download newlib fails around half of the time.  I will have it retry the `ct-ng build` command a couple times if it fails, but maybe you can think of a more robust solution.

aswaterman committed:
Retry ct-ng build because of occasional newlib download failure

aswaterman committed:
Speed up build a bit by disabling graphite and multilib

aswaterman committed:
Add toolchain to build path

aswaterman committed:
bump riscv-pk

aswaterman committed:
Hack up the toolchain paths

aswaterman committed:
fix PATH and add sanity check

aswaterman committed:
Fix openocd build by removing crosstools automake/aclocal

aswaterman commented:
@palmer-dabbelt it's getting further now... I needed to add a hack to copy some files around in the toolchain build, because out of the box it can't build executables.  (The C headers are in the wrong place, as are crt0.o and libm.a.)

aswaterman committed:
Put gcc in expected place for debug tests

aswaterman committed:
copy libc and libgloss, too

aswaterman committed:
fix gdb path

aswaterman commented:
OK... I finally got it sort of working, but I think this version of gdb doesn't interact well with this version of openocd, or maybe with the debug tests.  Not sure how we should proceed.

jim-wilson commented:
You need an FSF gdb with the RISC-V target description (aka xml) support.  This will appear in gdb-8.3 which hasn't been released yet.  Since it isn't a release yet, crosstools-ng doesn't use it.  Even with that pending gdb release, there will still be a few failures.  The python scripts in riscv-tests/debug need to be updated to match the new gdb output.  Gdb used to print a NaN if you have a float value in a 64-bit FP reg.  Now it does this
gdb) print $fa0
$1 = {float = 2, double = -nan(0xfffff40000000)}
(gdb) 
which is more useful, but riscv-tests/debug doesn't know how to parse that.  There may also be a couple of other misc failures.  The FP reg one if the main one that I remember.

We have a workaround for the gdb problem in riscv-gnu-toolchain mainline, and are using gdb mainline development instead of the last release.  I suppose it might be possible to add support to crosstools to do something similar, but someone would have to do some crosstools work to add an alternate gdb package for RISC-V.

aswaterman commented:
I propose we XFAIL the failing tests for now, rather than spending time to do something custom that we're going to throw away later.

aswaterman committed:
Expect 11 debug test failures

We need a GDB with RISC-V XML target description, which isn't in a release
yet, so isn't available through crosstools.

Additionally, the riscv-tests/debug scripts will need to be updated to
handle different GDB output when printing FP registers.

aswaterman commented:
Hooray, it passes under that assumption.

aswaterman committed:
Merge fesvr into spike repo; bump spike to v1.0.0

RobertHenry6bev commented:
Will this be merged in, or is it still baking?

aswaterman commented:
It's up to @palmer-dabbelt but I think it is sufficiently baked.  The remaining major issue with gdb and the debug tests can be fixed separately.

palmer-dabbelt commented:
A green checkmark is good enough for me :)
